src/native/containers/dn-malloc.h

```diff
@@ -0,0 +1,55 @@
+
+#ifndef __DN_MALLOC_H__
+#define __DN_MALLOC_H__
+
+#ifndef DN_USE_CUSTOM_MALLOC_H
+#include "dn-utils.h"
+
+#include <stdlib.h>
+#include <memory.h>
+
+static inline void
+dn_free (void *block)
+{
+	if (block != NULL)
+		free (block);
+}
+
+static inline void *
+dn_malloc (size_t size)
+{
+	if (!size)
+		return NULL;
```

Comment:
This is a bit odd. Typically if 0 is supplied, we should ensure an allocation is made a pass in a 1. If this logic is followed we fall into the classic case of not being able to distinguish from an allocation failure and an computed need of "0". The CoreCLR PAL handles this as follows:

https://github.com/dotnet/runtime/blob/153284a0605d5c07bdf8168573aa64afd3356beb/src/coreclr/pal/src/cruntime/malloc.cpp#L82-L106

Reply:
This is actually taken from mono's implementation of `g_malloc` that has the same behavior:

https://github.com/dotnet/runtime/blob/d87f7b2b81647917528286d91ed91db2a34b0985/src/mono/mono/eglib/gmem.c#L115

Searching through some documentations around malloc implementation you find things like this:

_If size is zero, the behavior of malloc is implementation-defined. For example, a null pointer may be returned. Alternatively, a non-null pointer may be returned; but such a pointer should not be [dereferenced](https://en.cppreference.com/w/c/language/operator_member_access), and should be passed to [free](https://en.cppreference.com/w/c/memory/free) to avoid memory leaks._

So I guess we could go both ways, but I have not problems to let malloc(0) do a 1 byte allocation if we believe we need to distinguish that from allocation failures returning NULL.

Reply:
Would it be better to assert that the passed in size is non-zero instead?

Reply:
> It isn't uncommon for some pointer arithmetic to be used as a "how much do I need?". This arithmetic can yield a 0 for empty length checks. 

Yes, and it is best for the collection classes to do not allocate anything in this case. We want the collection classes to be as efficient as possible.

Reply:
> We want the collection classes to be as efficient as possible.

Agree in principle. However, I would be shocked to see anyplace where this trick is noted in practice. This is an edge case that is better to sacrifice a incredibly small amount of impact for consistency. If we are going with "as efficient as possible" then we should just use statically allocated circular buffers and drop events like ETW. I don't think this zero allocation scenario warrants the "by all means necessary avoid allocations" argument.

Reply:
I agree that this is just a detail, and it does not have a large impact.

My underlying reason is that dislike wrappers of the classic C APIs like this. For example, are we going to have wrappers for memset or memcpy to avoid their undefined behaviors? My preference would be to avoid wrappers, call the C API directly and push the responsibility to avoid UBs to callers.

Reply:
> My underlying reason is that dislike wrappers of the classic C APIs like this.

We are in agreement here.

> For example, are we going to have wrappers for memset or memcpy to avoid their undefined behaviors?

Perhaps this is the confusion or maybe I am focusing on semantics here. This is implementation defined, not undefined. For the sake of the conversation let's clarify their differences. The implementation defined statement refers to implementing C compilers can define what they want and that is deemed "okay" because consumers can determine what is correct for any implementation. The notion of "undefined" is broader because it states that no implementing C compiler is permitted to change that because it is intrinsic to the language and C virtual machine specification. People may rely upon undefined behavior, but that is a risk and optimizers can do as they please. Sanitizers like ubsan focus on undefined behavior as opposed to implementation defined, which is permitted to be defined by the implementation.

Abstracting over undefined behavior is never okay and a fools errand, but providing an abstraction to be consistent across all implementation defined behavior does seem to have value.

If we go with the "no wrappers" approach, then we should be explicit about that policy and make a concerted effort to avoid them or abstract over them in a way that is consistent across the code base. I can't think of any other C API that warrants a wrapper other than `malloc`. I would be open to putting a stake in the ground that `malloc` is the only one we wrap if that would help move this forward.